specify/specify7
# Unhide string comparison operators in QueryBuilder

Fixes #1035

### Overview/Description

Adds a new Preference in the `QueryBuilder -> Behavior` section of User Preferences called `Show comparison operators for text-based fields` (default is unechecked/false)

<img width="1437" alt="new_query_pref" src="https://github.com/user-attachments/assets/c731e9d8-bed5-4b64-b52f-1eb80959f202" />

When checked, the preference unhides the following operators for text fields in the Query Builder:

```
>, Greater than
<, Less than
>=, Greater than or equal to
<=, Less than or equal to
```

All of these were able to be accomplished previously using the BETWEEN operator, but doing so required knowledge of how MySQL used the BETWEEN operator 'behind the scenes' (which itself would require knowledge of ASCII characters and their values). 

This preference can lead to potentially nonsensical results, especially when numeric strings are sorted lexicographically  (see https://github.com/specify/specify7/pull/3501#pullrequestreview-1431053184). 
This potential confusion is the primary motivation behind putting this functionality behind a preference and making it non-default. 

#### Edge Case Behaviors/Notes

_**Viewing existing queries which have comparison-based filters on a text field when the preference is disabled**_

If you do not have the preference checked but encounter an existing comparison filter on a text field, then Specify respects the filter on the field: leaving it untouched. However, if you then switch the filter to a non-comparison operator, then the previous comparison filter will be removed from the list of available filters; only reloading the page, enabling the preference, or generally leaving the page and coming back (i..e, reloading the Query) will allow the user. to"go back" to the previous filter.  

https://github.com/user-attachments/assets/8eeb034e-b4b1-4262-ae0a-6d64b6a673a4

Let me know if this edge case deserves more attention! A more user-friendly implementation for example might keep the previous comparison-based filter in the list of possible filters to choose from: until the query is saved or the web-page is left. 

_**Viewing queries which have comparison-based filters on a text field in older versions**_

Support for this feature is not retroactive: if such a filter exists on a query, the filter will be replaced with `Any` when viewing the query in older versions of Specify 7. 
When viewing the Query in Specify 6, Specify 6 will replace the filter with `Contains` (see https://github.com/specify/specify7/pull/3501#pullrequestreview-1431047000). The Query, however, should be hidden in Specify 6 by default (see https://github.com/specify/specify7/pull/3501#issuecomment-1553170095)

@specify/ux-testing @specify/dev-testing 
Feel free to give any feedback on the wording of the preference, the default value, etc.! 😄 

We can also consider defining this User Preference as a Remote Preference instead, if the scope of the preference should be broader than user (which might become important when/if [Shared Queries (#256)](https://github.com/specify/specify7/issues/256) are implemented)

### Checklist

      self-explanatory (or properly documented)

### Testing instructions

- With the `Show comparison operators for text-based fields` preference set to false (the default behavior)
- Open a new or existing query with a table containing a text field which is not CatalogNumber (e.g., `text1`, `text2`, `remarks`)
  - CatalogNumber should already contain the comparison operators: regardless of preference value 
  - If you do not have a text field in mind, https://files.specifysoftware.org/schema/version/2.10/ 
- Add the text field to the query if it is not already present
- Save the Query
- [ ] Click on/open the Filter Type dropdown (Any, Like, Equal, Between, etc.) and ensure that the Greater Than and Less Than filters are not shown
- Go into User Preferences and under `QueryBuilder -> Behavior` check/enable the `Show comparison operators for text-based fields` preference
  - Make sure the title and description of the preference makes sense!  
- Without reloading the page, go back to the previous Query and click on/open the Filter Type dropdown
- [ ] Ensure the Greater Than and Less Than filters are present in the dropdown
- Select one of the Greater Than or Less Than filters on a text field and save the Query
  - Optionally, ensure that results are returned as expected* when using comparison-based filters on text fields
- Navigate back to User Preferences and disable the `Show comparison operators for text-based fields` preference 
- [ ] Navigate back to the previously used Query and ensure that the comparison-based filter is present (and still respected by the query results) until changed
